lib/Dialect/TTNN/Transforms/Passes.cpp

```diff
@@ -3,20 +3,105 @@
 // SPDX-License-Identifier: Apache-2.0
 
 #include "ttmlir/Dialect/TTNN/Transforms/Passes.h"
-
+#include "mlir/Analysis/Liveness.h"
 #include "mlir/Dialect/Func/IR/FuncOps.h"
-#include "mlir/Dialect/MLProgram/IR/MLProgram.h"
-#include "mlir/Dialect/Tensor/IR/Tensor.h"
 #include "mlir/IR/PatternMatch.h"
-#include "mlir/Rewrite/FrozenRewritePatternSet.h"
-#include "mlir/Support/LogicalResult.h"
-#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
-#include "ttmlir/Conversion/TTIRToTTNN/TTIRToTTNN.h"
-#include "ttmlir/Dialect/TT/IR/TT.h"
-#include "ttmlir/Dialect/TT/IR/TTOpsTypes.h"
-#include "ttmlir/Dialect/TTIR/IR/TTIR.h"
-#include "ttmlir/Dialect/TTIR/IR/TTIROps.h"
-#include "ttmlir/Dialect/TTIR/Transforms/Passes.h"
 #include "ttmlir/Dialect/TTNN/IR/TTNNOpsTypes.h"
 
-namespace mlir::tt::ttnn {} // namespace mlir::tt::ttnn
+namespace mlir::tt::ttnn {
+#define GEN_PASS_DEF_TTNNDEALLOCATE
+#include "ttmlir/Dialect/TTNN/Transforms/Passes.h.inc"
+
+class TTNNDeallocate : public impl::TTNNDeallocateBase<TTNNDeallocate> {
+
+public:
+  using impl::TTNNDeallocateBase<TTNNDeallocate>::TTNNDeallocateBase;
+
+  Operation *getLastValueUsageOp(const LivenessBlockInfo *livenessInfo,
+                                 Value value) {
+    Operation *startOp = livenessInfo->getStartOperation(value);
+    Operation *endOp = livenessInfo->getEndOperation(value, startOp);
+    auto *opOperandIter =
+        llvm::find_if(endOp->getOpOperands(), [&](OpOperand &opOperand) {
+          return opOperand.is(value);
+        });
+
+    // Keep going until we find the last usage of the tensor.
```

Comment:
Maybe something like below to make while more readable?
It's fine with me either way, but it seems to me that break condition is more clear :D

```cpp
  // Keep going until we find the last usage of the tensor.
  // For DPS, stop if operand is not the DPS init one.
  // For non-DPS, skip through as it might modify the tensor in place.
  //
  while (opOperandIter != endOp->getOpOperands().end()) {
    bool isDpsInitOp = isa<DestinationStyleOpInterface>(endOp) &&
                       cast<DestinationStyleOpInterface>(endOp)
                           .isDpsInit(&(*opOperandIter));

    bool isNonDpsWithResults = !isa<DestinationStyleOpInterface>(endOp) &&
                               endOp->getNumResults() > 0;

    if (!(isDpsInitOp || isNonDpsWithResults))
      break;

    assert(endOp->getResults().size() == 1);
    OpResult result = endOp->getResult(0);
    endOp = livenessInfo->getEndOperation(result, endOp);

    opOperandIter =
        llvm::find_if(endOp->getOpOperands(), [&](OpOperand &opOperand) {
          return opOperand.is(result);
        });
  }
```

Reply:
I think I did write it with break statement at first and figured out I was dumb for doing so as while statement check is a break check by design. :)

Reply:
I'll think about improving readability of the code.